### PR TITLE
Adding warehouse class with ut

### PR DESCRIPTION
**Pull request by Aleksiiej**

### CODE REVIEW

File: src/app/medicine/medicine.hpp
Comment (BaqWin) [inline src/app/medicine/medicine.hpp:16]: Change class structure so private members are at the top of the class.

**Comment by ziobron** [comment]
Thanks @BaqWin for Code Review.
🏅 1 XP granted

File: src/app/warehouse/warehouse.hpp
Comment (BaqWin) [inline src/app/warehouse/warehouse.hpp:18]: Validate functions should be private.

File: src/app/warehouse/warehouse.hpp
Comment (BaqWin) [inline src/app/warehouse/warehouse.hpp:26]: Change class structure so private members are at the top of the class.
Add class extent.

File: src/app/medicine/medicine.cpp
Comment (BaqWin) [inline src/app/medicine/medicine.cpp:5]: Think about adding private function for expiration date validation. We don't want to add medicine that is already expired.

File: src/app/warehouse/warehouse.hpp
Comment (BaqWin) [inline src/app/warehouse/warehouse.hpp:25]: sed::set is better suited for our needs to store associations.

File: src/app/product/product.hpp
Comment (BaqWin) [reply src/app/product/product.hpp:19]: Consider adding pure virtual function that returns string with information about product.

File: src/app/product/product.hpp
Comment (BaqWin) [inline src/app/product/product.hpp:18]: Make this variable unsigned or change it to size_t.

File: src/app/product/product.hpp
Comment (BaqWin) [inline src/app/product/product.hpp:17]: Consider making setter for this variable with validation so it won't go below 0.

**Review by BaqWin — CHANGES_REQUESTED** [review]

File: src/app/warehouse/warehouse.cpp
Comment (mBialczak) [inline src/app/warehouse/warehouse.cpp:57]: Same consideration as in validateAdd().
I doubt it is needed but I miss it's intended use cases.

File: src/app/warehouse/warehouse.cpp
Comment (mBialczak) [inline src/app/warehouse/warehouse.cpp:62]: Just a suggestion: maybe getProduct() name is enough for consistensy with getProducts()?

File: src/app/warehouse/warehouse.hpp
Comment (mBialczak) [reply src/app/warehouse/warehouse.hpp:25]: std::set seems ritght at first glance, but might lead to storing more than one Product of the same type in the container (which might be a good idea itself- please read further remarks). std::set uses operator < for comparison but on addresses.

This leeds me to wonder if we even want to store shared_ptrs or objects.
Maybe storing Product object will be enough?
It depends mainly on the question if we assume existance of more than one Product object in container.
1. we can add Product by copying it or moving it our container.
2. we can return a raw pointer to our Product object stored in the vector or vector<*> in getProducts() case
3. we return nullptr from getProduct() if we don't find it anyway.
After all it seems, that a Warehouse will be the sole owner of The product.

In such case we could consider std::set again but it is not obvious for me if it is a good idea:
1. we would have to add operator< for Product (not a great problem after all)
2. we would have to check if product is already stored any way in order not to override an existing one (thus removing all data)

**Most important:** we should decide if we want to store **more objects of the same Product name**?
At first glance the answer is "no". But if we have for example same Medicine but with different expiration dates (one already stored may have shorter one than one added to the Warehouse). In such case we can not simply add them in one object.
I think that the answer to this question is **crucial** as it will impact implementation of many Warehouse member functions (if I analysed it right - currently we predict only one Product object stored).

**Review by mBialczak — COMMENTED** [review]
Lots of work done. Well done. But this set of classes is very compliated (Warehouse especially). Some thingit status might be polished. Some important design question arised.

**Comment by ziobron** [comment]
Thanks @mBialczak for Code Review.
🏅 1 XP granted

File: src/app/equipment/CMakeLists.txt
Comment (Aleksiiej) [reply src/app/equipment/CMakeLists.txt:3]: Done

File: src/app/equipment/equipment.cpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.cpp:1]: Properly filled file has been added.

File: src/app/equipment/equipment.hpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.hpp:6]: Proper constructor has been added.

File: src/app/medicine/medicine.cpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.cpp:5]: Function added with // TODO comment (there is no data type Date currently, so writing proper function is in this moment impossible, but will be possible in future).

File: src/app/medicine/medicine.hpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.hpp:16]: Done

File: src/app/product/product.cpp
Comment (Aleksiiej) [reply src/app/product/product.cpp:14]: amount has now uint type, so it won't go below 0.

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp:9]: Done

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp:12]: I assume function getName() is avalaible for Medicine and equipment classes even if is not virtual.

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp:17]: Done

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp:18]: Done.

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp:19]: Done

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp:30]: Done

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp:37]: I have created validateAdd() because this function was in UML diagram. I am opened for discussion

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp:57]: I have created validateAdd() because this function was in UML diagram. I am opened for discussion

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp:62]: Nack

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp:26]: Why class extent is needed here?

Comment (BaqWin) [reply src/app/warehouse/warehouse.hpp]: We want to make class extent in every class, no matter how many instances gonna be initialised. Right now we haven't made a decision if we'll have only one or few warehouses, no matter the count, we should always have static function that points to an instance of an object.

**Comment by ziobron** [comment]
Thanks @BaqWin for Code Review.
🏅 1 XP granted

File: src/app/product/product.hpp
Comment (Aleksiiej) [reply src/app/product/product.hpp]: Done

File: src/app/medicine/medicine.cpp
Comment (mBialczak) [inline src/app/medicine/medicine.cpp:15]: I think two best options exist for us, than returning vector<string>:
1. Stick to seperate getters (getExpirationDate(), getStorageTemperature etc.)
2. Return structure containing: expiration, storage temperature, vector<string> with compositions.

I don't know which is better for us, but both are easier to maintain and less error prone.

Here we stick to fixed string structure very dificult to parse and keep aligned with setter:
1. If we change anything in return we might broke a lot of code using getInfo()
2. Every time somebody calls getInfo, he/she/them/they (or whatever) will have to study structure of returned to string to extract a piece of interesting information. And can easily get confused.

File: src/app/product/product.cpp
Comment (mBialczak) [reply src/app/product/product.cpp:15]: I think that uint this does not protect us from anything and can give us a lot of BIGGER RISK.
Suppose you have amount == 2 at the start. And try to decerase the amount by 4.
The uint will wrap around storing new amount of 4294967295 or 4294967294 or whatever UINT_MAX -2 is.
Unsinged ints are very prone to underflow, overflow and stuff like that

**Review by mBialczak — COMMENTED** [review]
1. shit tests have some memory issues on valgrind.
2. Suggestion: it is usually good to rebase on top of current main branch state between review phases. I suppose you intended to do it after all comments are resolved, but the state between review and main might change a lot and it might require another round of reviewing.

**Comment by ziobron** [comment]
Thanks @mBialczak for Code Review.
🏅 1 XP granted

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp]: Done

File: src/app/medicine/medicine.cpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.cpp:15]: Because we use pointers to Product, those getters must be placed as virtual functions in Product class and also in Medicine class and it leads to creation of unnecessary code in those classes.

File: src/app/equipment/equipment.cpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.cpp:9]: Because we use pointers to Product, those getters must be placed as virtual functions in Product class and also in Medicine class and it leads to creation of unnecessary code in those classes.

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp]: Ok, I'm doing it with std::set

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp]: I will delete them

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp]: I will delete them

File: src/app/equipment/equipment.cpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.cpp:9]: Adjustment to std::map

File: src/app/medicine/medicine.cpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.cpp:15]: Adjustment to std::map

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp]: Done

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp]: Because we want store more objects of the same Product name - I choose to use std::vector instead of std::set.

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp]: Done

File: src/app/warehouse/warehouse.hpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.hpp]: Done

File: src/app/equipment/equipment.cpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.cpp]: Done

File: src/app/medicine/medicine.cpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.cpp]: Done

File: src/app/warehouse/warehouse.cpp
Comment (Aleksiiej) [reply src/app/warehouse/warehouse.cpp]: Done

File: src/app/equipment/equipment.cpp
Comment (BaqWin) [inline src/app/equipment/equipment.cpp:7]: Please use initializer list in constructor.

File: src/app/medicine/medicine.cpp
Comment (BaqWin) [inline src/app/medicine/medicine.cpp:8]: Please use initializer list in constructor

File: src/app/product/product.cpp
Comment (BaqWin) [inline src/app/product/product.cpp:6]: Please use initializer list in constructor

File: src/app/warehouse/warehouse.cpp
Comment (mBialczak) [inline src/app/warehouse/warehouse.cpp:36]: According to std::set::insert() documentation, insert "Inserts element(s) into the container, if the container doesn't already contain an element with an equivalent key."

So, after good change to set simple insert instruction below should be enough.

**Review by mBialczak — COMMENTED** [review]
Everything looks good in my opinion. 
Single comment is a minor one. Code could as well be left there as it is.
Anyway, before merging it would be good to rebase review ontop current main.
Maybe some conflicts will be there requiring resolving.
Besides, the pull request does not run checks present on main, so GH might not allow to merge any way.
This commit's history starting to get a little bit long. I think it is good enough to merge.
In general you have done lots of great job, so cograts!!
If anything arises future changes can be done.

**Comment by ziobron** [comment]
Thanks @mBialczak for Code Review.
🏅 1 XP granted

File: src/app/medicine/medicine.cpp
Comment (Aleksiiej) [reply src/app/medicine/medicine.cpp]: Done

File: src/app/product/product.cpp
Comment (Aleksiiej) [reply src/app/product/product.cpp]: Done

File: src/app/equipment/equipment.cpp
Comment (Aleksiiej) [reply src/app/equipment/equipment.cpp]: Done

**Review by mBialczak — APPROVED** [review]
Well done

**Comment by ziobron** [comment]
Thanks @mBialczak for Code Review.
🏅 1 XP granted

**Review by BaqWin — APPROVED** [review]

**Comment by ziobron** [comment]
Thanks @BaqWin for Code Review.
🏅 1 XP granted

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @Aleksiiej for making course materials better!

**Pull request merged**
